### PR TITLE
Add multiplayer, multiphase, or variable turn order support.

### PR DESCRIPTION
**Pull request by philihp**
## TL;DR:

* Added multiplayer/multiphase turn support during backpropagation
* Removed `Playout` object, replaced with shorter method - this is a breaking interface change, but the gem has a major version number of `0`, so that's acceptable.

## Details:

Adds a method to the game_state interface to ask it who the winner is, rather than if the current player won. Uses this during backpropagation, increasing the value of the node if the move was made by the acting player. The main use-case for this is with multiplayer games. For instance, with a game where players are red, blue, and green, the red player wants to maximize a move while minimizing two following moves by blue and green.

With multi-phase turns, for example we might have a game of blackjack, where a player can hit, hit, hit, and stay. Even though we don't wait for the dealer to act, each hit represents a decision made by the player. This could be generalized as the player choosing between 1, 2, 3, 4, or 5 hits before being forced to stay, however between some of those might be buying insurance or doubling down. It's simpler and probably better to model the game based on how the actual game works, and the state being an intermediate between each action/move.

Since we aren't just flipping the `won` boolean, variable turn order is also supported.

Additionally I don't think flipping the stats on the root node should be necessary if you select the best move based on which of the root's children has the most visits. I don't think it's even necessary to track the value of the root node; just its visits, which is used for UCT selection.

### CODE REVIEW

File: lib/mcts/node.rb [hunk 1]
```diff
@@ -40,43 +40,32 @@ def expand
...
-      end
+    def update_stats(winner:)
+      @visits += 1
+      @wins += 1 if winner == acting_player
     end
 
     private
 
+    def acting_player
+      return -1 if parent.nil?
```
Comment (PragTob) [inline lib/mcts/node.rb]: Why do we return `-1` here under these circumstances? Shouldn't we always be able to just return the last_turn_color? Plus, if we wanna exit early I believe `false/nil` are better values :)

File: lib/mcts/node.rb [hunk 2]
```diff
@@ -40,43 +40,32 @@ def expand
...
-    def backpropagate(won)
+    def backpropagate(winner:)
       node = self
-      node.update_won won
-      until node.root? do
-        won = !won # switching players perspective
-        node = node.parent
-        node.update_won(won)
-      end
+      node.update_stats(winner: winner) until (node = node.parent).nil?
```
Comment (PragTob) [inline lib/mcts/node.rb]: Any reasoning against using the `root?` method?

File: lib/mcts/playout.rb [hunk 3]
```diff
@@ -1,22 +0,0 @@
...
-      my_color = @game_state.last_turn_color
-      playout
-      @game_state.won?(my_color)
-    end
-
-    def playout
-      until @game_state.finished?
-        @game_state.set_move @game_state.generate_move
-      end
-    end
```
Comment (PragTob) [inline lib/mcts/playout.rb]: I see that your version is simpler, I'd still vote to keep the `Playout` object as itself alive. It's in the spirit of the Single Responsibility Principle (the node shouldn't know how to do this imo) + for benchmarking a playout can be benchmarked independently without needing to setup a node or tree. See `playout_help` :)

```
def playout_object_for(size)
  MCTS::Playout.new(Rubykon::GameState.new Rubykon::Game.new(size))
end
```

**Review by PragTob — COMMENTED** [review]
Hey, thanks for the PR!

Nice to get some multiplayer support :)

Lots of tests are failing now though, not sure where all of the failures come from - some come from the DoubleStep implementation others might come due to subtle changes.

Regarding tracking values in root, sure it's not strictly necessary. I found the value interesting to have handy to see what the general winning percentage is of all the traversed paths and it doesn't really have a high cost associated to it.

File: lib/mcts/node.rb (hunk 1 above)
Comment (philihp) [reply lib/mcts/node.rb]: Oh, i just wanted it to always return a number which would never == winner. I ran into an issue in a multi-player game where on the first move of the game there wasn't a previous last_turn_color.

I'm cool with returning `nil`; just a personal habit to avoid returning nil in languages without optionals :)

File: lib/mcts/node.rb (hunk 2 above)
Comment (philihp) [reply lib/mcts/node.rb]: Sounds good to me.

File: lib/mcts/node.rb [hunk 4]
```diff
@@ -40,43 +40,32 @@ def expand
...
-      @visits += 1
-      @wins += 1
-    end
-
-    def lost
-      @visits += 1
-    end
-
-    def backpropagate(won)
+    def backpropagate(winner:)
```
[the remark this answers is not in the document]
Comment (philihp) [reply lib/mcts/node.rb]: no problem :)

File: lib/mcts/playout.rb (hunk 3 above)
Comment (philihp) [reply lib/mcts/playout.rb]: Ah, I see - yeah this makes sense to me now.
